keps/sig-cli/2368-kubectl-default-container/README.md

```diff
@@ -0,0 +1,277 @@
+> supplied flags.  Today the commands are compiled into the client, which as the following challenges:
+
+Sidecars are popular now.
+Alternatively, a way to generically describe the default container for anything that happens to need a "default", which could also be used by external tooling.
+
+Issues https://github.com/kubernetes/kubernetes/issues/96986 opened by @howardjohn 
+> However, it gets worse because aside from a warning, this couples the default exec container to the container order. 
+> The container ordering also happens to have an impact on container startup order. We have started offering an option 
+> to inject our sidecar as the first container (previously, it was the last one), which has resulted in users running 
+> kubectl exec and getting the "wrong" container.
+
+
+### Goals
+
+
+- Provide a way to know which is the default container in pod for cli: kubectl
+
+
+### Non-Goals
+
+- If the cli is not kubectl, we don't determine which is the default container.
+
+
+## Proposal
+
+
+### Current Cli Behaviors
+
+Many Kubectl commands make requests to specific containers, `kubectl logs` and `kubectl exec` are most commonly
+used.
+
+There are currently 5 commands that allow specifying --container flag:
+- kubectl attach
+- kubectl cp
+- kubectl debug
+- kubectl exec
+- kubectl logs
+However, `kubectl debug` use this option to specify container name for debug container. 
+
+All commands that need to specify container name:
+
+- kubectl attach - Attach to a running container
+- kubectl cp - Copy files and directories to and from containers.
+- kubectl exec - Execute a command in a container
+
+These three commands above are similar. If omitted, the first container in the pod will be chosen.
+
+- kubectl logs - Print the logs for a container in a pod
+
+`kubectl logs` is a little different. Not default value for it. 
+However, it supports logs all containers with `--all-containers`.
+
+
+https://github.com/kubernetes/kubernetes/pull/87809 already add an annotation named
+`kubectl.kubernetes.io/default-logs-container` for log to use as default.
+
+We don't have a general default container name attribute for pod, and this would change pod spec and not acceptable.
+Why not use `kubectl.kubernetes.io/default-container` as the default container name annotation for all four commands
+above?
+
+### Proposal Details
+
+A single and generic annotation for all above commands like `kubectl.kubernetes.io/default-container` is a good choice to avoid needing many new annotations in the future.
+
+There are currently 5 commands that allow specifying --container flag, but not all of them consume this annotation.
+- `kubectl debug` will not consume this annotation.
+- kubectl exec
+- `kubectl logs` will consume this annotation only if no `--all-containers` option is specified.
+- kubectl attach
+- kubectl cp 
+
+As logs and exec are frequently used by users, we start from these two commands. `kubectl attach and cp` will be supported as well.
+
```

Comment:
I guess there's a constraint/caveat missing here, that as the annotation will not be automatically added, users and Pod owners will need to be aware of this new annotation.